### PR TITLE
Model Validation Infrastructure, Script, and Workflow

### LINKED ISSUE #404: Script to Automatically Test Featured Models
We need a script and potentially an action which (_e.g._, once per week or upon execution) performs unit tests on all of our featured models.

### PR DESCRIPTION
**Pull request by cwitkowitz**
Implements a model validation framework executed manually or as a daily GitHub Action. The main script headlessly drives the pyharp gradio endpoints (`/controls`, `/process`) across two tiers: local `pyharp/examples/` apps, and every Hugging Face Space under `teamup-tech`. 

Every model gets a default case with inputs synthesized from its own `/controls` spec; `config.yml` adds per-model and shared cases, declarative `expect` rules on outputs (audio/MIDI/label properties), and custom `validators`.

See `model_validation/README.md` for more details.

One current limitation is the lack of precision in tracking ZeroGPU quota usage. Since quota usage is dynamic and can be refunded, it might be worthwhile to retry ZeroGPU models before outright skipping the remaining ones after a single ZeroGPU quota exceeded error. Another workaround could be to set the number of workers to 1 to avoid eating up extra quota due to initial time reservations.

### CODE REVIEW

File: model_validation/src/cases.py [hunk 1]
```diff
@@ -0,0 +1,503 @@
...
+                missing[label] = (f"cannot synthesize input for "
+                                  f"file_types={spec.get('file_types')}; supply "
+                                  f"one via a test case's `files` entry")
+            args.append(handle_file(str(path)) if path is not None else None)
+        elif ctype in ("slider", "number_box"):
+            value = spec.get("value")
+            if value is None:
+                value = spec.get("minimum", 0)
+            args.append(value)
+        elif ctype == "text_box":
```
Comment (ChughVansh) [inline model_validation/src/cases.py:112]: Here, if a text_box's default value happens to be "", "" or "test" would resolve to "test." Since the intent is default values everywhere, this may be an edge case we're overlooking. 

elif ctype == "text_box":
    args.append(spec.get("value") if spec.get("value") is not None else "test")
OR
elif ctype == "text_box":
    value = spec.get("value")
    args.append(value if value is not None else "test")

**Review by 2cylu2 — COMMENTED** [review]
I ran the validation script locally against both local example servers (--local-examples) and remote Spaces:
--local-examples --load-only passes all 4 models.
Running full inference (--local-examples) fails for midi_synthesizer and ui_tester with:
AppError: The upstream Gradio app has raised an exception but has not enabled verbose error reporting. To enable, set show_error=True in launch().

It might be helpful to enable show_error=True either in pyharp's build_endpoint() or when launching the local example servers in harness.py. That way, validate_models.py can capture the actual traceback in report.json instead of the generic AppError, which should make debugging failures easier.

One thought on the ZeroGPU reservation issue mentioned in the PR description: instead of reducing --workers to 1 globally (which could slow down CPU model validation), we could run only the ZeroGPU models one at a time while keeping CPU models running in parallel. Since this script runs automatically, taking a little longer seems like a reasonable tradeoff if it helps avoid unnecessary quota usage.

One model (generalizable-accompaniment-generation) failed because it requires audio between 3 and 30 seconds long.
Would it make sense to increase the default synthesized audio duration to 3 or 5 seconds in assets.py, or should models with duration requirements always specify custom test files in config.yml under test_data/?

File: model_validation/src/cases.py (hunk 1 above)
Comment (cwitkowitz) [reply model_validation/src/cases.py:112]: Is there any situation where a text box would have `value=None`? We don't set a default in pyharp, but does it show up as None or "" if unspecified?

Comment (ChughVansh) [reply model_validation/src/cases.py:112]: I just checked with

> import gradio as gr
> tb = gr.Textbox()
> print(tb.value)

This returned None. But gr.Textbox("") will return "", so I'm assuming we get the latter. 

I looked through and i think the space has models deployed where either is the expected, so it may be good to test for both cases there? 

MMT, MuScriptor and STAGE have '' defaults
And there are several others that have None as a default.

**Review by 2cylu2 — APPROVED** [review]

File: model_validation/src/results.py [hunk 2]
```diff
@@ -98,33 +113,60 @@ def write_reports(results: list, out_dir: Path, label: str,
...
+    lines += ["| Model | Status | Stage | Hardware | Controls | Cases | Time (s) | Detail |",
+              "|---|---|---|---|---|---|---|---|"]
+
+    for r in ranked:
+        cases = ", ".join(f"{c.name} {case_emoji(c.ok)}" for c in r.cases) or "—"
         link = (f"[{r.target}](https://huggingface.co/spaces/{r.target})"
                 if r.kind == "space" else f"`{r.target}`")
-        detail = r.error.replace("|", "\\|")[:300] if r.error else ""
+        # Keep the table scannable; the untruncated text follows below
+        detail = r.error.replace("|", "\\|") if r.error else ""
```
Comment (ChughVansh) [inline model_validation/src/results.py:150]: replace("\n", " ") would probably be good. Omnivoice's error had a newline and kinda broke the whole table preview in the .md file.

File: model_validation/src/harness.py
Comment (ChughVansh) [inline model_validation/src/harness.py:66]: Recall the 'uncorrectable ecc error' i was telling you about. The one that can be solved by re-running the space. Turns out there's another one like that - "ReadTimeout: The read operation timed out." I faced this with two models today and re-running them gave the actual errors. What if along with loading_markers, we also had some a marker check setup that'd restart the space on particular input in errors?

FALSE_ERROR_MARKERS = ("read operation timed out", "uncorrectable ecc error").

+ an accompanying checker like is_loading_error()

and then it can either be in call_through_loading() as an 'or' in the if statement

however, if i'm understanding it right, connect_timeout's value is set to 420, which means it'll keep checking for 7 minutes. That should work, but for this one, we could lower the threshold? or limit by number of attempts? maybe have another if statement below is_loading_error():

if is_false_error(exc) and attempts < some_retry_limit:
                attempts += 1
                time.sleep(some_interval_value)

File: model_validation/src/harness.py [hunk 3]
```diff
@@ -426,22 +468,42 @@ def test_space(space_id: str, token: str, assets: Assets,
...
-        # SLEEPING/STARTING stage seen before the wake-up
+        # SLEEPING/STARTING stage seen before the wake-up, and fill in the
+        # hardware if it was not reported while the space was asleep
         if result.stage != "RUNNING":
             result.stage = "RUNNING"
+        if not result.hardware:
+            try:
+                awake = api.get_space_runtime(space_id)
+                result.hardware = awake.hardware or awake.requested_hardware or ""
+                zerogpu = is_zerogpu(result.hardware)
```
Comment (ChughVansh) [inline model_validation/src/harness.py:479]: shouldn't throttle be updated here in case the model was asleep earlier and uses cpu-basic? 

throttle = zerogpu or not result.hardware

Not a bug, but serializes a CPU model behind GPU models below if throttle is left in its stale state from line 427.

**Review by ChughVansh — CHANGES_REQUESTED** [review]
Catching a \n error in results.py to help with report.md formatting.

Also, an edge case catch and suggestion in harness.py.

Rest everything looks nice! Left more in-depth comments in the review

**Comment by ChughVansh** [comment]
Also, another question, is there any specific reason teamup-tech isn't treated as the default in --exclude or --spaces or even config.yml? doing so would be a very short code change and wouldn't require us to write 

--spaces teamup-tech/model1, teamup-tech/model2, teamup-tech/model3

and instead just have

--spaces model1, model2, model3

something like:
if opts.spaces:
    spaces = [s if "/" in s else f"{opts.org}/{s}" for s in opts.spaces]
    space_ids = [s for s in spaces if s not in excluded]

in line 203 @ validate_model.py. This would also allow us to input 

--spaces model1, model2, model3, diff-org/model-name

File: model_validation/src/harness.py
Comment (ChughVansh) [reply model_validation/src/harness.py:66]: Got another error today that was sorted by retrying

RemoteProtocolError: Server disconnected without sending a response.

File: model_validation/src/validate_models.py
Comment (ChughVansh) [inline model_validation/src/validate_models.py:341]: may be a good idea to add a validate_token(token) method here since right now if I pass in HF_TOKEN='random_mess' the script runs tests anyways. I was thinking this could be a method in utils.py using whoami() and error handling in there.

**Review by ChughVansh — APPROVED** [review]
Looks great! Just one thing. 

There are times when GPU exhaustion is read as a failure and not a skip; these were the two messages that triggered that:

1. The backend's ZeroGPU quota is exhausted for the identity making this call. Paste your own Hugging Face token in the token field (read scope) so usage is attributed to your account (OmniVoice, Melodyflow, etc.). Custom error handling from automatic model deployment agent.

2. You have exceeded your Pro ZeroGPU quota (180s requested vs. 111s left). Try again in 22:05:43. Add credits to keep using ZeroGPU. $1 for 10 minutes of ZeroGPU inference (DDSP Guitar). Regex matching for this format - 180s requested vs. 111s left - may be helpful bcs i've encountered this too

**Review by 2cylu2 — COMMENTED** [review]
I also think this looks great! I got the same messages as @ChughVansh.

For the first one, this was intentional with the two-Space approach I used for those models. Without a token, requests from the frontend Space are treated as anonymous calls, so Hugging Face counts them against the anonymous ZeroGPU quota even though the backend is functioning correctly. I put our HF token as a secret in those Spaces, but if the token is exhausted, users can enter their own token in the space provided. I might need to redeploy some of the models to add the option to enter a token, though, since that was a later addition.

I made the agent show this error when a Space doesn't have a secret HF token or when the token is exhausted. I manually added my token to some of the models that showed this message, and they worked. So it seems like the deployment itself is working, but this could be inconvenient for model validation since these currently show up as failures. It would probably make sense to treat these cases as skips instead. The regex suggestion for the Pro ZeroGPU quota message also makes sense.

**Comment by cwitkowitz** [comment]
Ok, to keep the reviewing and changes manageable, I'm going to merge this PR and open issues for:
1. Secondary (downstream) ZeroGPU errors trigger our skipping mechanism
2. Customizing MIDI instrument number for synthesized files
